### PR TITLE
Search/virtualize list

### LINKED ISSUE #23620: Virtualize dashboard search results list
**What would you like to be added**:
Improve performance of large dashboard lists via [react-window](https://github.com/bvaughn/react-window) and/or search results memoization. 

**Why is this needed**:
Current results list is a bit slow with over 1k dashboards.

### PR DESCRIPTION
**Pull request by Clarity-89**
**What this PR does / why we need it**:

Use react-window to virtualize search results list. Really noticeable performance increase for lists with about 1.5k dashboards. 

In order for this to work the height of the list items container has to be set dynamically, creating an extra scrollable container, which should be less of an issue with full screen dashboards search. 
Screenshot 2020-04-20 at 18 26 43
  

**Which issue(s) this PR fixes**:

Fixes #23620 

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by torkelo** [comment]
You should be able to fix the scrollbar issue, and use CustomScrollbar inside the react-window container

**Comment by torkelo** [comment]
with the folders and nesting it's a bit tricky, maybe we only virtualize results in search mode ? (and in search use a flat layout not nested in folders)

**Comment by Clarity-89** [comment]
> You should be able to fix the scrollbar issue, and use CustomScrollbar inside the react-window container

I actually had to remove the `CustomScrollbar` cause I can't get the proper `offsetTop`, most likely because it has a few nested divs with own positioning. But even if it worked, we'd still end up having two scrollbars since we do need to be able scroll both dashboards and folders.

> with the folders and nesting it's a bit tricky, maybe we only virtualize results in search mode ? (and in search use a flat layout not nested in folders)

This could work, also there would still be an issue with results being slow if there are lots of search hits (1k+).

In general I think this double scrollbar shouldn't be too big of an issue with full width search, the way it is in manage dashboards now.

**Comment by Clarity-89** [comment]
@torkelo I found a way to use the custom scrollbar, so it looks better now.

**Comment by torkelo** [comment]
Took this for a spin, sadly the double scrollbars makes it close to unsable. Many cases where I could not scroll up or down to see items in another expanded folder. 

I think we should be able to virtualize everything but then we would have to have a flat list where folders live as items and dashboards are indented. This nested structure is making this approach impossible.

**Comment by Clarity-89** [comment]
> I think we should be able to virtualize everything but then we would have to have a flat list where folders live as items and dashboards are indented. This nested structure is making this approach impossible.

A flat list would also work better for sorting, but I don't think we have time to make changes before 7.0 as it also requires some UI/UX input. 

I think this should be less of an issue for fullscreen search, similar to how this works in manage dashboard, but we can also apply virtualization only for the list view for now.

**Comment by torkelo** [comment]
> A flat list would also work better for sorting, but I don't think we have time to make changes before 7.0 as it also requires some UI/UX input. 

yes, agree. 

If possible try to only use virtualized only when searching and present list as flat (not nested)

**Comment by Clarity-89** [comment]
@torkelo now the results in the flat list view are virtualised. Could also look into virtualising dashboards for expanded folder but then we'd probably need to make the whole search overlay scrollable.

File: public/app/features/search/components/SectionHeader.tsx [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+    <div className={styles.wrapper} onClick={onSectionExpand}>
+      <SearchCheckbox editable={editable} checked={section.checked} onClick={onSectionChecked} />
+      <Icon className={styles.icon} name={section.icon as IconName} />
+
+      <span className={styles.text}>{section.title}</span>
+      {section.url && (
+        <a
+          href={section.url}
+          className={styles.link}
+          onClick={() => appEvents.emit(CoreEvents.hideDashSearch, { target: 'search-item' })}
```
Comment (torkelo) [inline public/app/features/search/components/SectionHeader.tsx:51]: what is this event for? is this needed / used now that we use a url query parm?

File: public/app/features/search/components/SearchResults.tsx [hunk 2]
```diff
@@ -25,43 +28,62 @@ export const SearchResults: FC<Props> = ({
...
+      </ul>
+    );
+  };
 
-    return section.items.map(item => (
-      <SearchItem key={item.id} {...{ item, editable, onToggleChecked, onTagSelected }} />
-    ));
+  const items = getVisibleItems(results);
+  const count = items.length;
+  const height = getItemsHeight(count, listHeight);
```
Comment (torkelo) [inline public/app/features/search/components/SearchResults.tsx:58]: is this function really needed, can't it always be same height as container?

File: public/app/features/search/components/SearchResults.tsx [hunk 3]
```diff
@@ -25,43 +28,62 @@ export const SearchResults: FC<Props> = ({
   onToggleChecked,
   onToggleSection,
   results,
+  wrapperRef,
   layout,
 }) => {
   const theme = useTheme();
   const styles = getSectionStyles(theme);
+  const listHeight = useListHeight(wrapperRef?.current?.offsetTop);
```
Comment (torkelo) [inline public/app/features/search/components/SearchResults.tsx:36]: Do we need this, in other places we have used AutoSizer to measure element height or width.

File: public/app/features/search/hooks/useListHeight.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+  useEffect(() => {
+    window.addEventListener('resize', handleResize);
+    return () => window.removeEventListener('resize', handleResize);
+  }, []);
+
+  useEffect(() => {
+    handleResize();
+  }, [offsetTop]);
+
+  return height;
```
Comment (torkelo) [inline public/app/features/search/hooks/useListHeight.ts:29]: think you can remove this and use AutoSizer instead to get dynamic height of the container  (disable width as width has extra checks)

File: public/app/features/search/components/SectionHeader.tsx (hunk 1 above)
Comment (Clarity-89) [reply public/app/features/search/components/SectionHeader.tsx:51]: Good catch, missed that one.

File: public/app/features/search/components/SearchResults.tsx (hunk 3 above)
Comment (Clarity-89) [reply public/app/features/search/components/SearchResults.tsx:36]: `AutoSizer`didn't work properly when virtualizing nested items, but sine we don't do that now Ill give it another try.

**Review by torkelo — APPROVED** [review]
Looks good! Great job, this should have a big impact on search performance for large installations & companies! :heart:

**Pull request merged**